File: leetcode/#151.py
# question --> https://leetcode.com/problems/reverse-words-in-a-string/

class Solution:
    def reverseWords(self, s: str) -> str:
        charArray = list(s.split())
        start, end = 0, len(charArray) - 1
        
        while start <= end: 
            charArray[start], charArray[end] = charArray[end], charArray[start] 
            start += 1
            end -= 1
            
        return ' '.join(charArray)
        
        # TC: O(N); SC: O(N)
        
# Alternative Approach

class Solution:
    def reverseWords(self, s: str) -> str:
        charArray = list(s.split())
        charArray = charArray[::-1]
        return ' '.join(charArray)
        
        # TC: O(N); SC: O(N)
        
# end 

File: leetcode/#151.js
// question --> https://leetcode.com/problems/reverse-words-in-a-string/

/**
 * @param {string} s
 * @return {string}
 */
var reverseWords = function(s) {
    let ans = new Array(); 
    let start = 0; 
    let end = 0; 
    let space = ' ';
    s = s.trim() + space; 
    let spaceFlag = false; 
    for (let index = 0; index < s.length; index++) {
        if ((s[index] !== space) && (!spaceFlag)) {
            start = index; 
            spaceFlag = true; 
        } else if (((s[index] === space) && (spaceFlag))) {
            end = index; 
            ans.push(s.slice(start, end));
            spaceFlag = false; 
        } 
    } 
    let i = 0; 
    let j = ans.length - 1; 
    while (i <= j) {
        [ans[i], ans[j]] = [ans[j], ans[i]];
        i += 1; 
        j -= 1;
    }
    return ans.join(' ');
 };
 
 // TC: O(N); SC: O(N)
